File: Agora-Rtm-Tutorial-macOS-Objective-C/Agora-Rtm-Tutorial/MainViewController.m
//
//  MainViewController.m
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/2/19.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "MainViewController.h"
#import "PeerChannelViewController.h"
#import "AgoraRtm.h"

@interface MainViewController () <PeerChannelVCDelegate>
@property (weak, nonatomic) IBOutlet NSTextField *accountTextField;
@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)viewWillAppear {
    [super viewWillAppear];
    [self logout];
}

- (void)prepareForSegue:(NSStoryboardSegue *)segue sender:(id)sender {
    NSString *identifier = segue.identifier;
    
    if (![identifier isEqualToString:@"mainToTab"]) {
        return;
    }
    
    PeerChannelViewController *vc = (PeerChannelViewController *)segue.destinationController;
    vc.delegate = self;
}

- (IBAction)doLoginPressed:(NSButton *)sender {
    [self login];
}

- (void)login {
    NSString *account = self.accountTextField.stringValue;
    if (!account.length) {
        return;
    }
    
    [AgoraRtm setCurrent:account];
    
    [AgoraRtm.kit loginByToken:nil user:account completion:^(AgoraRtmLoginErrorCode errorCode) {
        if (errorCode != AgoraRtmLoginErrorOk) {
            return;
        }
        
        [AgoraRtm setStatus:LoginStatusOnline];
        
        __weak MainViewController *weakSelf = self;
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf performSegueWithIdentifier:@"mainToTab" sender:nil];
        });
    }];
}

- (void)logout {
    if (AgoraRtm.status == LoginStatusOffline) {
        return;
    }
    
    [AgoraRtm.kit logoutWithCompletion:^(AgoraRtmLogoutErrorCode errorCode) {
        if (errorCode != AgoraRtmLogoutErrorOk) {
            return;
        }
        
        [AgoraRtm setStatus:LoginStatusOffline];
    }];
}

- (void)peerChannelVCWillClose:(PeerChannelViewController *)vc {
    vc.view.window.contentViewController = self;
}

@end

File: Agora-Rtm-Tutorial-macOS/Agora-Rtm-Tutorial/MainViewController.swift
//
//  MainViewController.swift
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/1/17.
//  Copyright © 2019 CavanSu. All rights reserved.
//

import Cocoa

class MainViewController: NSViewController {
    @IBOutlet weak var accountTextField: NSTextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear() {
        logout()
    }
    
    override func prepare(for segue: NSStoryboardSegue, sender: Any?) {
        guard let identifier = segue.identifier, identifier == "mainToTab" else {
            return
        }
        
        let peerChannelVC = segue.destinationController as! PeerChannelViewController
        peerChannelVC.delegate = self
    }
    
    @IBAction func doLoginPressed(_ sender: NSButton) {
        login()
    }
}

private extension MainViewController {
    func login() {
        let account = accountTextField.stringValue
        guard account.count > 0 else{
            return
        }
        
        AgoraRtm.current = account

        AgoraRtm.kit?.login(byToken: nil, user: account) { [unowned self] (errorCode) in
            guard errorCode == .ok else {
                return
            }
            
            AgoraRtm.status = .online
            
            DispatchQueue.main.async {
                self.performSegue(withIdentifier: "mainToTab", sender: nil)
            }
        }
    }
    
    func logout() {
        guard AgoraRtm.status == .online else {
            return
        }
        
        AgoraRtm.kit?.logout(completion: { (error) in
            guard error == .ok else {
                return
            }
            
            AgoraRtm.status = .offline
        })
    }
}

extension MainViewController: PeerChannelVCDelegate {
    func peerChannelVCWillClose(_ vc: PeerChannelViewController) {
        vc.view.window?.contentViewController = self
    }
}
